uniffi_bindgen/src/lib.rs

```diff
@@ -0,0 +1,267 @@
+) -> Result<()> {
+    // If --manifest-path is not provided, we just assume that the .idl
+    // file is in the `src` folder of the crate.
+    let manifest_path = match manifest_path_override {
+        Some(p) => PathBuf::from(p)
+            .canonicalize()
+            .map_err(|e| anyhow!("Failed to find Cargo.toml file: {:?}", e))?,
+        None => {
+            let src_dir = idl_file.parent().ok_or_else(|| anyhow!("no parent!"))?;
+            if src_dir.file_name().ok_or_else(|| anyhow!("not a folder"))? != "src" {
+                bail!("Could not determine the crate path from the .idl file");
+            }
+            let cargo_toml = src_dir
+                .parent()
+                .ok_or_else(|| anyhow!("no parent!"))?
+                .join("Cargo.toml");
+            if !cargo_toml.exists() {
+                bail!("Could not find Cargo.toml file")
+            }
+            cargo_toml
+        }
+    };
+    let metadata = cargo_metadata::MetadataCommand::new()
+        .manifest_path(manifest_path)
+        .exec()
+        .map_err(|e| anyhow!("Failed to run cargo metadata: {:?}", e))?;
+    let uniffi_runtime_deps: Vec<cargo_metadata::Package> = metadata
+        .packages
+        .into_iter()
+        .filter(|p| p.name == "uniffi")
+        .collect();
+    if uniffi_runtime_deps.is_empty() {
+        bail!("It looks like the crate doesn't depend the `uniffi` runtime. Please add `uniffi` as a dependency.");
+    }
+    if uniffi_runtime_deps.len() > 1 {
+        bail!("It looks like the workspace depends on multiple versions of `uniffi`. Please rectify the problem and try again.");
+    }
+    let uniffi_bindgen_version = BINDGEN_VERSION.parse::<u64>().expect("Should not fail");
+    let uniffi_runtime_version = uniffi_runtime_deps[0].version.major;
+    if uniffi_runtime_version != uniffi_bindgen_version {
+        bail!("The `uniffi` dependency version ({}) is different than `uniffi-bindgen` own version ({}). Please rectify the problem and try again.", uniffi_runtime_version, uniffi_bindgen_version);
+    }
+    Ok(())
+}
+
+// Generate the bindings in the target languages that call the scaffolding
+// Rust code.
+pub fn generate_bindings(
+    idl_file: &OsStr,
+    target_languages: Vec<&str>,
+    out_dir_override: Option<&OsStr>,
+) -> Result<()> {
+    let idl_file = PathBuf::from(idl_file)
+        .canonicalize()
+        .map_err(|e| anyhow!("Failed to find idl file: {:?}", e))?;
+    let component = parse_idl(&idl_file)?;
+    let out_dir = get_out_dir(&idl_file, out_dir_override)?;
+    for language in target_languages {
+        bindings::write_bindings(&component, &out_dir, language.try_into()?)?;
+    }
+    Ok(())
+}
+
+// Run tests against the foreign language bindings (generated and compiled at the same time).
+// Note that the cdylib we're testing against must be built already.
+pub fn run_tests(cdylib_dir: &OsStr, idl_file: &OsStr, test_scripts: Vec<&str>) -> Result<()> {
```

Comment:
The new test machinery living in `uniffi_bindgen` is here!

Reply:
From an API-consumer point of view, I'm lightly surprised to see this take the directory of the cdylib rather than the path to the cdylib itself. Not a big deal, just noting my surprise.